### ISSUE TITLE
Compare different approaches

### ISSUE THREAD

**Issue by gmpl** [polytypic/FsOptics#1]
Hi Vesa,

Long time ago I asked @mausch if we can [add Lenses to F#+](https://github.com/gmpl/FSharpPlus/issues/3) and he told me he was more interested in Polymorphic Lenses.

Last year I was researching this area and came up with a similar solution which I added to [F#+](https://github.com/gmpl/FSharpPlus/blob/06216d1ea557cdb6508568a5dc935a09c9347ed2/FSharpPlus/Lens.fs).

I need to analyse in detail what's the difference with your solution but I think you avoided to use a generic <code>map</code> by using a single Functor which combines both <code>ID</code> and <code>Const</code> whereas I used <code>map</code> and also the Profunctor abstraction from [FsControl](https://github.com/gmpl/FsControl) but later I found that from Profunctor the most used instance is the one for functions so I simplified it a bit.

I wonder if we can join forces and come up with a reasonable solution for this type of optics. What do you think?

**Comment by polytypic** [comment]
I'm happy to discuss ideas!  I'm not yet happy with the results (in this repository) and my understanding of the topic.

I think you articulated one of the basic ideas exploited here correctly.  The other idea was to eliminate higher-rank polymorphism using mutation.  The idea has been to try and see whether types that are not well supported in F# could be eliminated using alternative approaches.  It should be possible to do a more or less "faithful" encoding of the Haskell approach (e.g. [Higher](https://github.com/palladin/Higher/blob/master/src/Higher.Core/FTLens.fs)), but it becomes quite complex and possibly expensive (object allocations and virtual functions).

**Comment by gmpl** [comment]
I agree that it will be nice to eliminate or at least reduce the higher order types simulated through member constraints. If you take the approach of Higher you would move the overhead from compile-time to run-time, if you ask me I would prefer to keep it at compile time.
Regarding the compile-time performance I think the F# compiler still has room to improve it but also noted that if you use the <code>|></code> operator it helps type inference a lot.
The other problem I faced was that the Lens Haskell library is huge and I wanted to keep it simple.
Another problem still not solved is how to abstract over folds while keeping it optimal, I mean not having to traverse the whole structure if it's not necessary, which in Haskell happens naturally.
